blackjax/adaptation/pathfinder_adaptation.py

```diff
@@ -0,0 +1,137 @@
+"""Implementation of the Stan warmup for the HMC family of sampling algorithms."""
+from typing import Any, Callable, List, NamedTuple, Tuple
+
+import jax
+import jax.numpy as jnp
+
+from blackjax.vi.pathfinder import (
+    init as pathfinder_init_fn,
+    sample_from_state,
+    lbfgs_inverse_hessian_formula_1
+    )
+
+from blackjax.adaptation.step_size import (
+    DualAveragingAdaptationState,
+    dual_averaging_adaptation,
+)
+from blackjax.mcmc.hmc import HMCState
+from blackjax.types import Array, PRNGKey
+
+__all__ = ["base"]
+
+
+class PathfinderAdaptationState(NamedTuple):
+    da_state: DualAveragingAdaptationState
+    inverse_mass_matrix: Array
+
+
+def base(
+    kernel_factory: Callable,
+    logprob_fn: Callable,
+    target_acceptance_rate: float = 0.65,
+):
+    """Warmup scheme for sampling procedures based on euclidean manifold HMC.
+    This function tunes the values of the step size and the mass matrix according 
+    to this schema:
+        * pathfinder algorithm is run and an estimation of the inverse mass matrix
+          is derived, as well as an initialization point for the markov chain
+        * Nesterov's dual averaging adaptation is then run to tune the step size
+
+    Parameters
+    ----------
+    kernel_factory
+        A function which returns a transition kernel given a step size and a
+        mass matrix.
+   logprob_fn
+        The log density probability density function from which we wish to sample.
+    target_acceptance_rate:
+        The target acceptance rate for the step size adaptation.
+
+    Returns
+    -------
+    init
+        Function that initializes the warmup.
+    update
+        Function that moves the warmup one step.
+    final
+        Function that returns the step size and mass matrix given a warmup state.
+
+    """
+    da_init, da_update, da_final = dual_averaging_adaptation(target=target_acceptance_rate)
+
+    def init(
+        rng_key: PRNGKey,
+        initial_position: Array,
+        initial_step_size: float
+    ) -> PathfinderAdaptationState:
+        """Initialize the warmup.
+
+        To initialize the warmup we use pathfinder to estimate the inverse mass matrix and
+        then we set up the dual averaging adaptation algorithm
+        """
+        da_state = da_init(initial_step_size)
+
+        pathfinder_rng_key, sample_rng_key = jax.random.split(rng_key, 2)
+        pathfinder_state = pathfinder_init_fn(
+                pathfinder_rng_key,
+                logprob_fn,
+                initial_position)
+        new_initial_position = sample_from_state(
+                sample_rng_key,
+                pathfinder_state,
+                1)[0]
```

Comment:
I understand, not liking it too. I tried to implement your idea but I couldn't due to the need to support pytrees as positions. Basically in `sample_from_state` we ravel the pytree and then map the unravel function over that leading axis. If we drop the leading axis in `lbfgs_sample` then we would need to implement a different logic based on `M` being 1 or >1. That would do more harm then good I think.

This made me realize that that `[0]` is actually wrong and will raise an error if `pathfinder_state.x` is a pytree instead of an array. Hence I've change the code into calling `jax.map_pyteee`. I think it's a bit more clear now but still hackish

Reply:
I see. In that case could you add a comment explaining why we need indexing here?

Reply:
Done